Tools/SetupUnityPackage/UnityUtils/FileHelper.cs

```diff
@@ -33,16 +35,35 @@ public static string GetInternalStorage()
 
             // On Android persistentDataPath returns the external folder, where the app may not have permissions to write.
             // we use reflection to call File.getAbsolutePath(currentActivity.getFilesDir())
-            using (var unityPlayer = new AndroidJavaClass("com.unity3d.player.UnityPlayer"))
-            using (var currentActivity = unityPlayer.GetStatic<AndroidJavaObject>("currentActivity"))
+            var path = string.Empty;
+
+            try
             {
-                var getFilesDir = AndroidJNIHelper.GetMethodID(AndroidJNI.FindClass("android/content/ContextWrapper"), "getFilesDir", "()Ljava/io/File;");
-                var internalFilesDir = AndroidJNI.CallObjectMethod(currentActivity.GetRawObject(), getFilesDir, Array.Empty<jvalue>());
-                var getAbsolutePath = AndroidJNIHelper.GetMethodID(AndroidJNI.FindClass("java/io/File"), "getAbsolutePath", "()Ljava/lang/String;");
+                if (AndroidJNI.AttachCurrentThread() == 0)
+                {
+                    using (var unityPlayer = new AndroidJavaClass("com.unity3d.player.UnityPlayer"))
+                    using (var currentActivity = unityPlayer.GetStatic<AndroidJavaObject>("currentActivity"))
+                    {
+                        var getFilesDir = AndroidJNIHelper.GetMethodID(AndroidJNI.FindClass("android/content/ContextWrapper"), "getFilesDir", "()Ljava/io/File;");
+                        var internalFilesDir = AndroidJNI.CallObjectMethod(currentActivity.GetRawObject(), getFilesDir, Array.Empty<jvalue>());
+                        var getAbsolutePath = AndroidJNIHelper.GetMethodID(AndroidJNI.FindClass("java/io/File"), "getAbsolutePath", "()Ljava/lang/String;");
+
+                        path = AndroidJNI.CallStringMethod(internalFilesDir, getAbsolutePath, Array.Empty<jvalue>());
+
+                        // if not on the unity thread
+                        if (Thread.CurrentThread.ManagedThreadId != 1)
```

Comment:
I know this is a weak way of checking if the code is running on the Unity thread. The value 1 was hand checked and consistently worked. May it change in future release? Unlikely, nevertheless it could happen.
Naming the Unity thread could be a good idea, but that thread can be named only within Unity's Monobehaviours in the scene. This is tricky for us to enforce.
If you have a suggestion, I'll look into it.